### ISSUE TITLE
Flash MoBA consumes more memory than PyTorch’s SDPA

### ISSUE THREAD

**Issue by JingangQu** [mit-han-lab/flash-moba#7]
Hi,

Congratulations on the great work. I’m very interested in Flash MoBA and have been running some tests on it. My main goal was to evaluate the impact of `moba_chunk_size` and `moba_topk`*on GPU memory usage and computation time, and to compare the results against PyTorch’s `scaled_dot_product_attention`.

I focused on the non-causal (`causal=False`) case and ran experiments on an H100 with:

* `batch_size = 1`
* `nheads = 4`
* `headdim = 128`
* `causal = False`

The test settings were:

```python
seqlens = [1024, 2048, 4096, 8192, 16384, 32768, 65536, 131072, 262144, 524288, 1_000_000]
chunk_sizes = [64, 128, 256, 512, 1024]
topks = [2, 4, 8, 16, 32]
```

The results are summarized as follows:

1. For a fixed sequence length, `moba_chunk_size` and `moba_topk` appear to have no impact on peak memory usage.
2. For a fixed sequence length, **larger** `moba_chunk_size` and `moba_topk` lead to longer computation time.
3. Flash MoBA is faster than SDPA.
4. However, Flash MoBA **consumes more memory** than SDPA.

<img width="4754" height="1750" alt="Image" src="https://github.com/user-attachments/assets/de41786c-00fb-4119-8ac0-a65ae1cdab63" />

The test code is as follows:

[test_moba_memory_benchmark.py](https://github.com/user-attachments/files/23728176/test_moba_memory_benchmark.py)

In addition, could you please help clarify the following questions:

1. In the original [[MoBA](https://github.com/MoonshotAI/MoBA/blob/master/moba/moba_efficient.py)](https://github.com/MoonshotAI/MoBA/blob/master/moba/moba_efficient.py) implementation, the final output is obtained by combining sparse attention and self-attention via online softmax — where self-attention performs local attention within each chunk (each token attends to previous tokens inside its own chunk), and sparse attention computes top-k cross-chunk attention (selected tokens attend to the top-k most relevant chunks).
   For Flash MoBA, does it only include the top-k cross-chunk attention, without the local self-attention component?

2. When `seqlen_k` is not divisible by `moba_chunk_size`, how is the tail chunk handled?

Thank you very much!

**Comment by JerryGJX** [comment]
Thank you for running these experiments and sharing the memory profile — this information is very valuable. I’ll take a closer look at it soon.

Regarding your questions:

1. Local self-attention support
We do support the local self-attention component. The implementation can be found here:
https://github.com/mit-han-lab/flash-moba/blob/39d9ac043b271d046a2181a9991e99a26b67bca1/csrc/flash_topk/src/flash_topk_kernel.h#L203-L214

   In short, during the top-k selection process, we explicitly ensure that the self-attention block is always selected for causal attention.

2. Mean pooling behavior
Mean pooling is applied only to the valid portion of the last block. The relevant implementation is here:
https://github.com/mit-han-lab/flash-moba/blob/39d9ac043b271d046a2181a9991e99a26b67bca1/flash_moba/triton_mean_pool.py#L95-L102

**Comment by JingangQu** [comment]
Thanks for your responses !

I really want to apply flash moba into my workflow. It is so promising !  However, greater memory consumption is really problematic.

I really appreciate it if you could solve it.


Great thanks !!!
